ads-api/src/main/java/io/voodoo/apps/ads/api/AdClientArbitrageur.kt

```diff
@@ -157,28 +160,61 @@ class AdClientArbitrageur(
     suspend fun fetchAdIfNecessary(
         localExtrasProvider: () -> Array<Pair<String, Any>>
     ): List<Result<Ad>?> = supervisorScope {
-        val localExtras by lazy(LazyThreadSafetyMode.PUBLICATION) { localExtrasProvider() }
+        val localExtras = lazy(LazyThreadSafetyMode.PUBLICATION) { localExtrasProvider() }
 
         clients
             .map { client ->
                 async {
-                    val availableAdCount = client.getAvailableAdCount().notLocked
-                    if (
-                        availableAdCount < requiredAvailableAdCount &&
-                        (enforceRequiredAvailableAdCount || requiredAvailableAdCount <= client.config.adCacheSize)
-                    ) {
-                        runCatching { client.fetchAd(*localExtras) }
-                            // client will throw AdAlreadyLoadingException if already loading
-                            .takeIf { it.exceptionOrNull() !is AdAlreadyLoadingException }
-                    } else {
-                        // No fetch needed
-                        null
-                    }
+                    fetchAdIfNecessary(client, localExtras)
                 }
             }
             .awaitAll()
     }
 
+    private suspend fun fetchAdIfNecessary(
+        client: AdClient<*>,
+        localExtrasProvider: Lazy<Array<Pair<String, Any>>>
+    ): Result<Ad>? {
+        val availableAdCount = client.getAvailableAdCount().notLocked
+        if (availableAdCount >= requiredAvailableAdCount) {
+            return null
+        }
+        if (
+            !overrideClientCacheSize &&
+            client.config.adCacheSize <= requiredAvailableAdCount &&
+            availableAdCount >= client.config.adCacheSize
+        ) {
+            return null
+        }
+
+        // Backoff
+        val lastFailInfo = lastFailInfoByClientMap[client]
+        if (backoffConfig != null && lastFailInfo != null) {
+            val delayForNextAttemptMillis =
+                backoffConfig.getDelay(lastFailInfo.attempt + 1)
+                    ?.inWholeMilliseconds
+                    ?: 0
+            val timeSinceLastAttempt = System.currentTimeMillis() - lastFailInfo.timestampMillis
```

Comment:
ok then it will be triggered on the next "polling" in our case, we don't need to use a delay here it assumes we will call it later again 👍

Reply:
yes